### DISCUSSION TITLE
Composed target-typed inference for switch expression and deconstruction

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [dotnet/csharplang#2898] [General]
Currently, this code compiles:
```csharp
MemberInfo info = null;
Action<object> getValue = null;

getValue = info switch
{
    PropertyInfo pi => () => pi.GetValue(d),
    FieldInfo fi => () => fi.GetValue(d)
};
```

This code also compiles:
```csharp
PropertyInfo pi = null;
(type, getValue) = (pi.PropertyType, () => pi.GetValue(d));
```

However, when combing switch expression and tuple together, this code cannot compile:
```csharp
MemberInfo info = null;
Action<object> getValue = null;
Type type = null;

(type, getValue) = info switch
{
    PropertyInfo pi => (pi.PropertyType, () => pi.GetValue(d)),
    FieldInfo fi => (fi.FieldType, () => fi.GetValue(d))
};
```

The compiler cannot infer type of lambda.
I'm not sure whether this belongs to target-typed expressions or common types. Maybe related to #2823.

**Comment by alrz** [comment]
This is a part of https://github.com/dotnet/csharplang/issues/1394. Per https://github.com/dotnet/csharplang/issues/1394#issuecomment-531013585 it includes switch expressions as well.
